src/object_store/src/object/mod.rs

```diff
@@ -933,5 +945,167 @@ pub async fn build_remote_object_store(
+    }
+}
+
+fn get_attempt_timeout_by_type(config: &ObjectStoreConfig, operation_type: OperationType) -> u64 {
+    match operation_type {
+        OperationType::Upload => config.retry.upload_attempt_timeout_ms,
+        OperationType::StreamingUploadInit
+        | OperationType::StreamingUpload
+        | OperationType::StreamingUploadFinish => config.retry.streaming_upload_attempt_timeout_ms,
+        OperationType::Read => config.retry.read_attempt_timeout_ms,
+        OperationType::StreamingReadInit | OperationType::StreamingRead => {
+            config.retry.streaming_read_attempt_timeout_ms
+        }
+        OperationType::Metadata => config.retry.metadata_attempt_timeout_ms,
+        OperationType::Delete => config.retry.delete_attempt_timeout_ms,
+        OperationType::DeleteObjects => config.retry.delete_attempt_timeout_ms,
+        OperationType::List => config.retry.list_attempt_timeout_ms,
+    }
+}
+
+struct RetryCondition {
+    operation_type: OperationType,
+    retry_count: usize,
+    metrics: Arc<ObjectStoreMetrics>,
+}
+
+impl RetryCondition {
+    fn new(operation_type: OperationType, metrics: Arc<ObjectStoreMetrics>) -> Self {
+        Self {
+            operation_type,
+            retry_count: 0,
+            metrics,
+        }
+    }
+
+    #[inline(always)]
+    fn should_retry_inner(&mut self, err: &ObjectError) -> bool {
+        let should_retry = err.should_retry();
+        if should_retry {
+            self.retry_count += 1;
+        }
+
+        should_retry
+    }
+}
+
+impl tokio_retry::Condition<ObjectError> for RetryCondition {
+    fn should_retry(&mut self, err: &ObjectError) -> bool {
+        self.should_retry_inner(err)
+    }
+}
+
+impl Drop for RetryCondition {
+    fn drop(&mut self) {
+        if self.retry_count > 0 {
+            self.metrics
+                .request_retry_count
+                .with_label_values(&[self.operation_type.as_str()])
+                .inc_by(self.retry_count as _);
+        }
+    }
+}
+
+async fn retry_request<F, T, B>(
+    builder: B,
+    config: &ObjectStoreConfig,
+    operation_type: OperationType,
+    object_store_metrics: Arc<ObjectStoreMetrics>,
+) -> ObjectResult<T>
+where
+    B: Fn() -> F,
+    F: Future<Output = ObjectResult<T>>,
+{
+    let backoff = get_retry_strategy(config, operation_type);
+    let timeout_duration =
+        Duration::from_millis(get_attempt_timeout_by_type(config, operation_type));
+    let operation_type_str = operation_type.as_str();
+
+    let retry_condition = RetryCondition::new(operation_type, object_store_metrics);
+
+    let f = || async {
+        let future = builder();
+        if timeout_duration.is_zero() {
+            future.await
+        } else {
+            tokio::time::timeout(timeout_duration, future)
+                .await
+                .unwrap_or_else(|_| {
+                    Err(ObjectError::timeout(format!(
+                        "{}_attempt_timeout_ms {:?} {}_retry_attempts {:?}",
```

Comment:
"retry attempts exhausted for {}. Please modify {}_attempt_timeout_ms (current=...) and {}_retry_attempts (current=...) under [storage.object_store.retry] in the config accordingly if needed."